### DISCUSSION TITLE
Close method removed of Pipeliner  in V9.  Don't I have to call it? Or through other means?

### DISCUSSION THREAD

**Discussion by hexxv** [redis/go-redis#2583] [Q&A] [answered]
V8: 

type Pipeliner interface {
	StatefulCmdable
	Len() int
	Do(ctx context.Context, args ...interface{}) *Cmd
	Process(ctx context.Context, cmd Cmder) error
	Close() error
	Discard() error
	Exec(ctx context.Context) ([]Cmder, error)
}

V9 :

type Pipeliner interface {
	StatefulCmdable

	// Len is to obtain the number of commands in the pipeline that have not yet been executed.
	Len() int

	// Do is an API for executing any command.
	// If a certain Redis command is not yet supported, you can use Do to execute it.
	Do(ctx context.Context, args ...interface{}) *Cmd

	// Process is to put the commands to be executed into the pipeline buffer.
	Process(ctx context.Context, cmd Cmder) error

	// Discard is to discard all commands in the cache that have not yet been executed.
	Discard()

	// Exec is to send all the commands buffered in the pipeline to the redis-server.
	Exec(ctx context.Context) ([]Cmder, error)
}

**Comment by monkey92t** [comment, marked as answer]
Yes, the pipeline does not require a "close" method, so we removed it in version 9. It does not cause any resource leaks.

**Comment by hexxv** [comment]
thx
